tests/functional/u/used/used_before_assignment_typing.py

```diff
@@ -180,6 +192,10 @@ def defined_in_nested_if_else(self) -> heapq:
         print(heapq)
         return heapq
 
+    def defined_in_type_checking(self) -> posix:  # [used-before-assignment]
+        print(pickle)  # [used-before-assignment]
+        return posix
+
```

Comment:
This test passes for me on main (and the title of the method doesn't tell me what it's testing).

Reply:
This tests for false negative, both names are defined in the TYPE_CHECKING block and should raise an error.

Reply:
But they do raise an error already without this PR, is what I'm saying. And the total number of errors goes from 3 down to 2, which isn't an improvement.

Apply this diff to main, and it passes:
```diff
diff --git a/tests/functional/u/used/used_before_assignment_ternary.txt b/tests/functional/u/used/used_before_assignment_ternary.txt
index d991970e4..bcb21df1c 100644
--- a/tests/functional/u/used/used_before_assignment_ternary.txt
+++ b/tests/functional/u/used/used_before_assignment_ternary.txt
@@ -1,3 +1,3 @@
-used-before-assignment:6:14:6:15:invalid:Using variable 'a' before assignment:HIGH
-used-before-assignment:8:14:8:15:invalid:Using variable 'b' before assignment:HIGH
-used-before-assignment:9:10:9:11:invalid:Using variable 'c' before assignment:HIGH
+used-before-assignment:6:14:6:15:invalid:Using variable 'a' before assignment:HIGH
+used-before-assignment:8:14:8:15:invalid:Using variable 'b' before assignment:HIGH
+used-before-assignment:9:10:9:11:invalid:Using variable 'c' before assignment:HIGH
diff --git a/tests/functional/u/used/used_before_assignment_typing.py b/tests/functional/u/used/used_before_assignment_typing.py
index a685bdabc..7f464f061 100644
--- a/tests/functional/u/used/used_before_assignment_typing.py
+++ b/tests/functional/u/used/used_before_assignment_typing.py
@@ -7,6 +7,8 @@ from typing import List, Optional, TYPE_CHECKING
 if TYPE_CHECKING:
     if True:  # pylint: disable=using-constant-test
         import math
+    elif input():
+        import pickle
     from urllib.request import urlopen
     import array
     import base64
@@ -195,3 +197,7 @@ class TypeCheckingMultiBranch:  # pylint: disable=too-few-public-methods,unused-
     def defined_in_with(self) -> base64:
         print(binascii)
         return base64
+
+    def defined_in_type_checking(self) -> posix:  # [undefined-variable]
+        print(pickle)  # [used-before-assignment]
+        return posix  # [undefined-variable]
diff --git a/tests/functional/u/used/used_before_assignment_typing.txt b/tests/functional/u/used/used_before_assignment_typing.txt
index c0a31fae0..54e15e0cb 100644
--- a/tests/functional/u/used/used_before_assignment_typing.txt
+++ b/tests/functional/u/used/used_before_assignment_typing.txt
@@ -1,5 +1,8 @@
-undefined-variable:68:21:68:28:MyClass.incorrect_typing_method:Undefined variable 'MyClass':UNDEFINED
-undefined-variable:73:26:73:33:MyClass.incorrect_nested_typing_method:Undefined variable 'MyClass':UNDEFINED
-undefined-variable:78:20:78:27:MyClass.incorrect_default_method:Undefined variable 'MyClass':UNDEFINED
-used-before-assignment:139:35:139:39:MyFourthClass.is_close:Using variable 'math' before assignment:HIGH
-used-before-assignment:152:20:152:28:VariableAnnotationsGuardedByTypeChecking:Using variable 'datetime' before assignment:HIGH
+undefined-variable:70:21:70:28:MyClass.incorrect_typing_method:Undefined variable 'MyClass':UNDEFINED
+undefined-variable:75:26:75:33:MyClass.incorrect_nested_typing_method:Undefined variable 'MyClass':UNDEFINED
+undefined-variable:80:20:80:27:MyClass.incorrect_default_method:Undefined variable 'MyClass':UNDEFINED
+used-before-assignment:141:35:141:39:MyFourthClass.is_close:Using variable 'math' before assignment:HIGH
+used-before-assignment:154:20:154:28:VariableAnnotationsGuardedByTypeChecking:Using variable 'datetime' before assignment:HIGH
+undefined-variable:201:42:201:47:TypeCheckingMultiBranch.defined_in_type_checking:Undefined variable 'posix':UNDEFINED
+used-before-assignment:202:14:202:20:TypeCheckingMultiBranch.defined_in_type_checking:Using variable 'pickle' before assignment:HIGH
+undefined-variable:203:15:203:20:TypeCheckingMultiBranch.defined_in_type_checking:Undefined variable 'posix':UNDEFINED
```

Reply:
Oh, I have an outdated version, this doesn't use `posix` anymore.

Reply:
My bad, I was changing the names to try and resolve some CI failures. Not very sure why the tests are failing on pypy, could be because of the unsafe implementation of `is_node_in_typing_guarded_import_block`.

Reply:
I understand this one better now. I think this is a good to leave on this PR, since I can't get the current version of this test to pass on main 👍